### ISSUE TITLE
detect CPU features at run time, not during compilation

### ISSUE THREAD

**Issue by mys721tx** [torognes/vsearch#342]
Currently VSEARCH detects the CPU features support at both compile time and runtime.
https://github.com/torognes/vsearch/blob/67a88bc444df7499ef0bb3fe33cb8d631a12d64b/src/cpu.cc#L107-L173

https://github.com/torognes/vsearch/blob/67a88bc444df7499ef0bb3fe33cb8d631a12d64b/src/vsearch.cc#L305-L335

This leads to serval problems:

1. Depending on the compiling machine, either `increment_counters_from_bitmap_ssse3 ` or `increment_counters_from_bitmap_sse2` will be generated at compile time.
2. SSSE3 is not enabled by default on some operating systems. (#327)
3. The addition of `-march=native` breaks binary compatibility for different microarchitectures. (#341)

Since there is also a need of distributing pre-compiled binaries, I propose that all CPU feature detection should move from compile time to runtime. In this way, all necessary instructions are generated and the machine that runs VSEARCH decides which function to use. In this way, maximum compatibility and efficiency can be ensured.

**Comment by torognes** [comment]
On the x86_64 architecture, the cpu.cc file is actually compiled twice, with and without SSSE3 defined. Both the `increment_counters_from_bitmap_ssse3` function and the `increment_counters_from_bitmap_sse2` function are therefore compiled. The function actually used is selected at run time.

I will reverse the change in #327 and make sure the distributed binaries are compatible with the broadest range of computers.

Users who want to get the best performance from their local machines may change the Makefile to add the `-march=native` option and compile their own binaries.

**Comment by mys721tx** [comment]
With`CFLAGS='-g'` and `CXXFLAGS='-g -std=c++11 -stdlib=libc++'` (an unrelated issue with clang on macOS 10.14), I compiled from 2168bb7. The following are the reversed compiled `increment_counters_from_bitmap_ssse3`:
```Assembly
        ; ================ B E G I N N I N G   O F   P R O C E D U R E ================


                     __Z36increment_counters_from_bitmap_ssse3PtPhj:        // increment_counters_from_bitmap_ssse3(unsigned short*, unsigned char*, unsigned int)
000000010003ace0         push       rbp                                         ; Begin of unwind block (FDE at 0x10004cf9c), CODE XREF=__Z16search_topscoresP12searchinfo_s+150
000000010003ace1         mov        rbp, rsp
000000010003ace4         add        edx, 0xf
000000010003ace7         shr        edx, 0x4
000000010003acea         je         loc_10003ad43

000000010003acec         movdqa     xmm0, xmmword [__ZL23minwordmatches_defaults+64] ; 0x100046b40
000000010003acf4         movdqa     xmm1, xmmword [__ZL23minwordmatches_defaults+80] ; 0x100046b50
000000010003acfc         pcmpeqd    xmm2, xmm2

                     loc_10003ad00:
000000010003ad00         movdqu     xmm3, xmmword [rsi]                         ; CODE XREF=__Z36increment_counters_from_bitmap_ssse3PtPhj+97
000000010003ad04         add        rsi, 0x2
000000010003ad08         pshufb     xmm3, xmm0
000000010003ad0d         por        xmm3, xmm1
000000010003ad11         pcmpeqb    xmm3, xmm2
000000010003ad15         movdqa     xmm4, xmm3
000000010003ad19         punpcklbw  xmm4, xmm4
000000010003ad1d         punpckhbw  xmm3, xmm3
000000010003ad21         movdqa     xmm5, xmmword [rdi]
000000010003ad25         psubsw     xmm5, xmm4
000000010003ad29         movdqa     xmm4, xmmword [rdi+0x10]
000000010003ad2e         psubsw     xmm4, xmm3
000000010003ad32         movdqa     xmmword [rdi], xmm5
000000010003ad36         movdqa     xmmword [rdi+0x10], xmm4
000000010003ad3b         add        rdi, 0x20
000000010003ad3f         dec        edx
000000010003ad41         jne        loc_10003ad00

                     loc_10003ad43:
000000010003ad43         pop        rbp                                         ; CODE XREF=__Z36increment_counters_from_bitmap_ssse3PtPhj+10
000000010003ad44         ret
                        ; endp
```

and `increment_counters_from_bitmap_sse2`:
```Assembly
        ; ================ B E G I N N I N G   O F   P R O C E D U R E ================


                     __Z35increment_counters_from_bitmap_sse2PtPhj:        // increment_counters_from_bitmap_sse2(unsigned short*, unsigned char*, unsigned int)
000000010003ad50         push       rbp                                         ; Begin of unwind block (FDE at 0x10004cfdc), CODE XREF=__Z16search_topscoresP12searchinfo_s+337
000000010003ad51         mov        rbp, rsp
000000010003ad54         add        edx, 0xf
000000010003ad57         shr        edx, 0x4
000000010003ad5a         je         loc_10003adb3

000000010003ad5c         movdqa     xmm0, xmmword [__ZL23minwordmatches_defaults+64] ; 0x100046b40
000000010003ad64         movdqa     xmm1, xmmword [__ZL23minwordmatches_defaults+80] ; 0x100046b50
000000010003ad6c         pcmpeqd    xmm2, xmm2

                     loc_10003ad70:
000000010003ad70         movdqu     xmm3, xmmword [rsi]                         ; CODE XREF=__Z35increment_counters_from_bitmap_sse2PtPhj+97
000000010003ad74         add        rsi, 0x2
000000010003ad78         pshufb     xmm3, xmm0
000000010003ad7d         por        xmm3, xmm1
000000010003ad81         pcmpeqb    xmm3, xmm2
000000010003ad85         movdqa     xmm4, xmm3
000000010003ad89         punpcklbw  xmm4, xmm4
000000010003ad8d         punpckhbw  xmm3, xmm3
000000010003ad91         movdqa     xmm5, xmmword [rdi]
000000010003ad95         psubsw     xmm5, xmm4
000000010003ad99         movdqa     xmm4, xmmword [rdi+0x10]
000000010003ad9e         psubsw     xmm4, xmm3
000000010003ada2         movdqa     xmmword [rdi], xmm5
000000010003ada6         movdqa     xmmword [rdi+0x10], xmm4
000000010003adab         add        rdi, 0x20
000000010003adaf         dec        edx
000000010003adb1         jne        loc_10003ad70

                     loc_10003adb3:
000000010003adb3         pop        rbp                                         ; CODE XREF=__Z35increment_counters_from_bitmap_sse2PtPhj+10
000000010003adb4         ret
                        ; endp
```

The same is repeated on a Linux box with GCC 8.2.1:
```Assembly
000000000003f580 <_Z36increment_counters_from_bitmap_ssse3PtPhj>:
   3f580:	83 c2 0f             	add    edx,0xf
   3f583:	f7 c2 f0 ff ff ff    	test   edx,0xfffffff0
   3f589:	74 69                	je     3f5f4 <_Z36increment_counters_from_bitmap_ssse3PtPhj+0x74>
   3f58b:	c1 ea 03             	shr    edx,0x3
   3f58e:	83 e2 fe             	and    edx,0xfffffffe
   3f591:	31 c0                	xor    eax,eax
   3f593:	66 0f 6f 05 15 c4 00 	movdqa xmm0,XMMWORD PTR [rip+0xc415]        # 4b9b0 <_ZL23minwordmatches_defaults+0x7540>
   3f59a:	00 
   3f59b:	66 0f 6f 0d 1d c4 00 	movdqa xmm1,XMMWORD PTR [rip+0xc41d]        # 4b9c0 <_ZL23minwordmatches_defaults+0x7550>
   3f5a2:	00 
   3f5a3:	66 0f 76 d2          	pcmpeqd xmm2,xmm2
   3f5a7:	66 0f 1f 84 00 00 00 	nop    WORD PTR [rax+rax*1+0x0]
   3f5ae:	00 00 
   3f5b0:	f3 0f 6f 1c 06       	movdqu xmm3,XMMWORD PTR [rsi+rax*1]
   3f5b5:	66 0f 38 00 d8       	pshufb xmm3,xmm0
   3f5ba:	66 0f eb d9          	por    xmm3,xmm1
   3f5be:	66 0f 74 da          	pcmpeqb xmm3,xmm2
   3f5c2:	66 0f 6f e3          	movdqa xmm4,xmm3
   3f5c6:	66 0f 60 e3          	punpcklbw xmm4,xmm3
   3f5ca:	66 0f 68 db          	punpckhbw xmm3,xmm3
   3f5ce:	66 0f 6f 2f          	movdqa xmm5,XMMWORD PTR [rdi]
   3f5d2:	66 0f e9 ec          	psubsw xmm5,xmm4
   3f5d6:	66 0f 6f 67 10       	movdqa xmm4,XMMWORD PTR [rdi+0x10]
   3f5db:	66 0f e9 e3          	psubsw xmm4,xmm3
   3f5df:	66 0f 7f 2f          	movdqa XMMWORD PTR [rdi],xmm5
   3f5e3:	66 0f 7f 67 10       	movdqa XMMWORD PTR [rdi+0x10],xmm4
   3f5e8:	48 83 c0 02          	add    rax,0x2
   3f5ec:	48 83 c7 20          	add    rdi,0x20
   3f5f0:	39 c2                	cmp    edx,eax
   3f5f2:	75 bc                	jne    3f5b0 <_Z36increment_counters_from_bitmap_ssse3PtPhj+0x30>
   3f5f4:	c3                   	ret    
   3f5f5:	66 2e 0f 1f 84 00 00 	nop    WORD PTR cs:[rax+rax*1+0x0]
   3f5fc:	00 00 00 
   3f5ff:	90                   	nop
```

```Assembly
000000000003f600 <_Z35increment_counters_from_bitmap_sse2PtPhj>:
   3f600:	83 c2 0f             	add    edx,0xf
   3f603:	f7 c2 f0 ff ff ff    	test   edx,0xfffffff0
   3f609:	74 62                	je     3f66d <_Z35increment_counters_from_bitmap_sse2PtPhj+0x6d>
   3f60b:	c1 ea 03             	shr    edx,0x3
   3f60e:	83 e2 fe             	and    edx,0xfffffffe
   3f611:	31 c0                	xor    eax,eax
   3f613:	66 0f 6f 05 a5 c3 00 	movdqa xmm0,XMMWORD PTR [rip+0xc3a5]        # 4b9c0 <_ZL23minwordmatches_defaults+0x7550>
   3f61a:	00 
   3f61b:	66 0f 76 c9          	pcmpeqd xmm1,xmm1
   3f61f:	90                   	nop
   3f620:	f3 0f 6f 14 06       	movdqu xmm2,XMMWORD PTR [rsi+rax*1]
   3f625:	66 0f 60 d2          	punpcklbw xmm2,xmm2
   3f629:	f2 0f 70 d2 50       	pshuflw xmm2,xmm2,0x50
   3f62e:	66 0f 70 d2 50       	pshufd xmm2,xmm2,0x50
   3f633:	66 0f eb d0          	por    xmm2,xmm0
   3f637:	66 0f 74 d1          	pcmpeqb xmm2,xmm1
   3f63b:	66 0f 6f da          	movdqa xmm3,xmm2
   3f63f:	66 0f 60 da          	punpcklbw xmm3,xmm2
   3f643:	66 0f 68 d2          	punpckhbw xmm2,xmm2
   3f647:	66 0f 6f 27          	movdqa xmm4,XMMWORD PTR [rdi]
   3f64b:	66 0f e9 e3          	psubsw xmm4,xmm3
   3f64f:	66 0f 6f 5f 10       	movdqa xmm3,XMMWORD PTR [rdi+0x10]
   3f654:	66 0f e9 da          	psubsw xmm3,xmm2
   3f658:	66 0f 7f 27          	movdqa XMMWORD PTR [rdi],xmm4
   3f65c:	66 0f 7f 5f 10       	movdqa XMMWORD PTR [rdi+0x10],xmm3
   3f661:	48 83 c0 02          	add    rax,0x2
   3f665:	48 83 c7 20          	add    rdi,0x20
   3f669:	39 c2                	cmp    edx,eax
   3f66b:	75 b3                	jne    3f620 <_Z35increment_counters_from_bitmap_sse2PtPhj+0x20>
   3f66d:	c3                   	ret    
   3f66e:	66 90                	xchg   ax,ax
```

With current setup, the SSSE3 instructions are generated for both functions on macOS.

**Comment by frederic-mahe** [comment]
Hi, I took the liberty to shorten the issue title. Feel free to revert.

**Comment by torognes** [comment]
I have changed compiler options to `-march=x86-64` and `-mtune=generic` to produce code that will run on all x86_64 systems while tuning for more modern cpus. This is available in VSEARCH version 2.9.1 just released.

**Comment by mys721tx** [comment]
On macOS 10.14 with 8be92ca the two functions still are the same.

```Assembly

        ; ================ B E G I N N I N G   O F   P R O C E D U R E ================


                     __Z36increment_counters_from_bitmap_ssse3PtPhj:        // increment_counters_from_bitmap_ssse3(unsigned short*, unsigned char*, unsigned int)
000000010003bd00         push       rbp                                         ; Begin of unwind block (FDE at 0x10004df9c), CODE XREF=__Z16search_topscoresP12searchinfo_s+150
000000010003bd01         mov        rbp, rsp
000000010003bd04         add        edx, 0xf
000000010003bd07         shr        edx, 0x4
000000010003bd0a         je         loc_10003bd64

000000010003bd0c         movdqa     xmm0, xmmword [__ZL23minwordmatches_defaults+64] ; 0x100047b40
000000010003bd14         movdqa     xmm1, xmmword [__ZL23minwordmatches_defaults+80] ; 0x100047b50
000000010003bd1c         pcmpeqd    xmm2, xmm2

                     loc_10003bd20:
000000010003bd20         movdqu     xmm3, xmmword [rsi]                         ; CODE XREF=__Z36increment_counters_from_bitmap_ssse3PtPhj+98
000000010003bd24         add        rsi, 0x2
000000010003bd28         pshufb     xmm3, xmm0
000000010003bd2d         por        xmm3, xmm1
000000010003bd31         pcmpeqb    xmm3, xmm2
000000010003bd35         movdqa     xmm4, xmm3
000000010003bd39         punpcklbw  xmm4, xmm4
000000010003bd3d         punpckhbw  xmm3, xmm3
000000010003bd41         movdqa     xmm5, xmmword [rdi]
000000010003bd45         psubsw     xmm5, xmm4
000000010003bd49         movdqa     xmm4, xmmword [rdi+0x10]
000000010003bd4e         psubsw     xmm4, xmm3
000000010003bd52         movdqa     xmmword [rdi], xmm5
000000010003bd56         movdqa     xmmword [rdi+0x10], xmm4
000000010003bd5b         add        rdi, 0x20
000000010003bd5f         add        edx, 0xffffffff
000000010003bd62         jne        loc_10003bd20

                     loc_10003bd64:
000000010003bd64         pop        rbp                                         ; CODE XREF=__Z36increment_counters_from_bitmap_ssse3PtPhj+10
000000010003bd65         ret
                        ; endp
```
```Assembly
        ; ================ B E G I N N I N G   O F   P R O C E D U R E ================


                     __Z35increment_counters_from_bitmap_sse2PtPhj:        // increment_counters_from_bitmap_sse2(unsigned short*, unsigned char*, unsigned int)
000000010003bd70         push       rbp                                         ; Begin of unwind block (FDE at 0x10004dfdc), CODE XREF=__Z16search_topscoresP12searchinfo_s+339
000000010003bd71         mov        rbp, rsp
000000010003bd74         add        edx, 0xf
000000010003bd77         shr        edx, 0x4
000000010003bd7a         je         loc_10003bddd

000000010003bd7c         movdqa     xmm0, xmmword [__ZL23minwordmatches_defaults+80] ; 0x100047b50
000000010003bd84         pcmpeqd    xmm1, xmm1
000000010003bd88         nop        dword [rax+rax]

                     loc_10003bd90:
000000010003bd90         movdqu     xmm2, xmmword [rsi]                         ; CODE XREF=__Z35increment_counters_from_bitmap_sse2PtPhj+107
000000010003bd94         add        rsi, 0x2
000000010003bd98         punpcklbw  xmm2, xmm2
000000010003bd9c         pshuflw    xmm2, xmm2, 0x50
000000010003bda1         pshufd     xmm2, xmm2, 0x50
000000010003bda6         por        xmm2, xmm0
000000010003bdaa         pcmpeqb    xmm2, xmm1
000000010003bdae         movdqa     xmm3, xmm2
000000010003bdb2         punpcklbw  xmm3, xmm3
000000010003bdb6         punpckhbw  xmm2, xmm2
000000010003bdba         movdqa     xmm4, xmmword [rdi]
000000010003bdbe         psubsw     xmm4, xmm3
000000010003bdc2         movdqa     xmm3, xmmword [rdi+0x10]
000000010003bdc7         psubsw     xmm3, xmm2
000000010003bdcb         movdqa     xmmword [rdi], xmm4
000000010003bdcf         movdqa     xmmword [rdi+0x10], xmm3
000000010003bdd4         add        rdi, 0x20
000000010003bdd8         add        edx, 0xffffffff
000000010003bddb         jne        loc_10003bd90

                     loc_10003bddd:
000000010003bddd         pop        rbp                                         ; CODE XREF=__Z35increment_counters_from_bitmap_sse2PtPhj+10
000000010003bdde         ret
                        ; endp
000000010003bddf         db  0x90 ; '.'                                         ; End of unwind block (FDE at 0x10004dfdc)
```
